lib/auth/trustedcluster.go

```diff
@@ -45,129 +45,115 @@ import (
 )
 
 // UpsertTrustedCluster creates or toggles a Trusted Cluster relationship.
-func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.TrustedCluster) (newTrustedCluster types.TrustedCluster, returnErr error) {
+func (a *Server) UpsertTrustedCluster(ctx context.Context, tc types.TrustedCluster) (newTrustedCluster types.TrustedCluster, returnErr error) {
+	// verify that trusted cluster role map does not reference non-existent roles
+	if err := a.checkLocalRoles(ctx, tc.GetRoleMap()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	// It is recommended to omit trusted cluster name because the trusted cluster name
 	// is updated to the roots cluster name during the handshake with the root cluster.
 	var existingCluster types.TrustedCluster
-	if trustedCluster.GetName() != "" {
+	var cas []types.CertAuthority
+	if tc.GetName() != "" {
 		var err error
-		existingCluster, err = a.GetTrustedCluster(ctx, trustedCluster.GetName())
+		existingCluster, err = a.GetTrustedCluster(ctx, tc.GetName())
 		if err != nil && !trace.IsNotFound(err) {
 			return nil, trace.Wrap(err)
 		}
 	}
 
-	enable := trustedCluster.GetEnabled()
-
-	// If the trusted cluster already exists in the backend, make sure it's a
-	// valid state change client is trying to make.
-	if existingCluster != nil {
-		if err := existingCluster.CanChangeStateTo(trustedCluster); err != nil {
-			return nil, trace.Wrap(err)
-		}
+	// if there is no existing cluster, switch to the create case
+	if existingCluster == nil {
+		return a.createTrustedCluster(ctx, tc)
```

Comment:
Concurrent calls to `UpsertTrustedCluster` can result in an `AlreadyExists` error, which at the very least feels at least a bit wrong.

Are we already retrying every operation in the terraform provider no matter what the error is? Will we remember to do so for `UpsertTrustedCluster` if and when that comes to be automatable?

Reply:
So I tried adding retry to this logic, but ended up discarding it.  Once you start to try to distinguish between `AlreadyExists` due to concurrent tc creation vs `AlreadyExists` due to conflicting CAs and handling `CompareFailed` from failed updates, it all starts to get a bit messy.  My (potentially weak) justification for leaving as-is is that anything currently updating these in a racy fashion is already broken because it is likely to cause conflicting state (e.g. role map on CA disagreeing with role map in TC).  Unexpected error seems like a strictly better outcome.  I can take another stab at if you feel strongly about it, but right now I'm leaning toward just accepting that this method might yield unexpected errors during concurrent use.

Reply:
If you find yourself around the code maybe we should replace the AlreadyExists with a CompareFailed here in UpsertTrustedCluster? Otherwise it's not even worth the mental effort in opening a PR, I agree.